apiserver/common/secrets/secrets.go

```diff
@@ -672,87 +672,136 @@ func RemoveSecretsForAgent(
+func removeSecretFromExternal(removeState SecretsRemoveState, adminConfigGetter BackendAdminConfigGetter, authTag names.Tag, uri *coresecrets.URI, revisions ...int) error {
+	var revs []*coresecrets.SecretRevisionMetadata
+	var err error
+	if len(revisions) == 0 {
+		// Remove all revisions.
+		revs, err = removeState.ListSecretRevisions(uri)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		for _, revId := range revs.RevisionIDs() {
-			if err = backend.DeleteContent(context.TODO(), revId); err != nil {
+	} else {
+		// Remove specified revisions.
+		revs = make([]*coresecrets.SecretRevisionMetadata, len(revisions))
+		for i, rev := range revisions {
+			revMeta, err := removeState.GetSecretRevision(uri, rev)
+			if err != nil {
 				return errors.Trace(err)
 			}
+			revs[i] = revMeta
 		}
-		if err := p.CleanupSecrets(&cfg, authTag, revs); err != nil {
-			return errors.Trace(err)
-		}
-		return nil
 	}
 
-	result := params.ErrorResults{
-		Results: make([]params.ErrorResult, len(args.Args)),
-	}
+	providersToCleanUp := make(map[string]provider.SecretRevisions)
 
-	if len(args.Args) == 0 {
-		return result, nil
-	}
-	cfgInfo, err := adminConfigGetter()
-	if err != nil {
-		return result, errors.Trace(err)
-	}
+	for _, rev := range revs {
+		backendId := rev.ValueRef.BackendID
+		revisionId := rev.ValueRef.RevisionID
 
-	removeFromExternal := func(uri *coresecrets.URI, revisions ...int) error {
-		externalRevs := make(map[string]provider.SecretRevisions)
-		gatherExternalRevs := func(valRef *coresecrets.ValueRef) {
-			if valRef == nil {
-				// Internal secret, nothing to do here.
-				return
-			}
-			if _, ok := externalRevs[valRef.BackendID]; !ok {
-				externalRevs[valRef.BackendID] = provider.SecretRevisions{}
-			}
-			externalRevs[valRef.BackendID].Add(uri, valRef.RevisionID)
-		}
-		if len(revisions) == 0 {
-			// Remove all revisions.
-			revs, err := removeState.ListSecretRevisions(uri)
+		// Repeatedly attempt to delete the revision from the backend until one of:
+		// * deletion is successful
+		// * deletion is unsuccessful with error==NotFound but the revision has not been moved to a new backend
+		// * deletion is unsuccessful with error!=NotFound
+		for {
+			cfgInfo, err := adminConfigGetter()
 			if err != nil {
 				return errors.Trace(err)
 			}
-			for _, rev := range revs {
-				gatherExternalRevs(rev.ValueRef)
-			}
-		} else {
-			for _, rev := range revisions {
-				revMeta, err := removeState.GetSecretRevision(uri, rev)
-				if err != nil {
-					return errors.Trace(err)
-				}
-				gatherExternalRevs(revMeta.ValueRef)
+			backendCfg := cfgInfo.Configs[backendId]
+			err = removeRevisionFromBackend(&backendCfg, revisionId)
+			if err == nil {
+				// Continue to the next revision if successful.
+				break
+			} else if !errors.Is(err, errors.NotFound) {
+				// Exit early for any error other than NotFound
+				return errors.Trace(err)
```

Comment:
We should Annotate to include the secret and revision that could not be deleted.

Reply:
Added some annotations everywhere in place of `errors.Trace`, although not sure I got the style right